Speed up paletized texture

PatrickvL opened the pull request:
Include a palette hash in the texture lookup key, instead of patching  Palette_Lock functions and freeing textures when setting palette's.

To avoid overflowing the resource cache, PruneResourceCache now applies cache eviction as well.
Thanks to that change, two `Palette_Lock` patches could be removed as well.

This speeds up titles that use paletized textures a lot, here `Yu-Gi-Oh! The Dawn Of Destiny` (YuGiOh DOD) running in-game at 30fps :
![image](https://user-images.githubusercontent.com/461930/69357262-16bead00-0c85-11ea-8912-65182c4ecc2f.png)

JusicP commented:
Half-Life 2 Crashes on startup.
```
[0x32F8] MAIN: Received Exception (Code := 0xC0000005)

 EIP := 0x04108AA0
 EFL := 0x00010246
 EAX := 0x00000100 EBX := 0x00000400 ECX := 0x00000000 EDX := 0xFFFFFFFF
 ESI := 0x00000000 EDI := 0x00000000 ESP := 0x099FF9D0 EBP := 0x099FF9DC
 CR2 := 0x00000000
```

PatrickvL commented:
Try the latest build (I force-pushed a safe-guard against that crash)

JusicP commented:
Now it works

literalmente-game commented:
These don't go in-game with the PR:

- MechAssault
- Metal Slug 5
- Crash N Burn

NZJenkins commented:
> These don't go in-game with the PR:
> 
> * MechAssault
> * Metal Slug 5
> * Crash N Burn
> 
> These have new rendering issues:
> 
> * Kill Switch (wrong textures being loaded for some models)
> * RalliSport Challenge (black floor)
> * Bloodrayne (black textures for many models, also black floor)
> * Bloodrayne 2 (black textures for most of the models, black floor)

Mechassault appears broken since 6b08ee81cc15b5d3c220 so may need to double check with the previous build

literalmente-game commented:
> > These don't go in-game with the PR:
> > 
> > * MechAssault
> > * Metal Slug 5
> > * Crash N Burn
> > 
> > These have new rendering issues:
> > 
> > * Kill Switch (wrong textures being loaded for some models)
> > * RalliSport Challenge (black floor)
> > * Bloodrayne (black textures for many models, also black floor)
> > * Bloodrayne 2 (black textures for most of the models, black floor)
> 
> Mechassault appears broken since [6b08ee8](https://github.com/Cxbx-Reloaded/Cxbx-Reloaded/commit/6b08ee81cc15b5d3c2209857deddcf5ffc07287d) so may need to double check with the previous build

Updated the comment, you're right, this list of games was affected by the mentioned change, not this PR:
- MechAssault
- Metal Slug 5
- Crash N Burn

AzurikRiseOfPerathia commented:
Azurik : Rise Of Perathia : Black texture, and some texture poorly modeled (Image has the support)

![2019-11-23_005825](https://user-images.githubusercontent.com/36118972/69468442-73f15600-0d8c-11ea-8cf8-9329d42fc809.png)

hyperspeedgx commented:
Does we need to set some kinda of settings for Yu-Gi-Oh works? My only shows a black screen, I saw that pal ver. works but I don't have it.

PatrickvL commented:
@AzurikRiseOfPerathia the exploding vertices are not related to this pull request, since it doesn't touch vertices at all.

It's probably caused by the recent MOVA fix, @NZJenkins is already looking into it.

As for the black textures, I'm investigating the root cause, but it's eluding me for now...

PatrickvL commented:
@literalmente-game commit db45305 may fix black textures, could you do a quick test?

PatrickvL commented:
Although the latest commit might improve/fix a few regressions, the bigger problem this PR still has, is that when the resource cache is pruned (which right now means it's cleared entirely), the association between Xbox render targets and the host counterparts is lost.

This results in the creation of new host-side render-targets, but those won't contain the same data as before anymore.

Also, some resources (like the backbuffer and other globals) should not be removed at all, but are also cleared right now. For this, we have to come up with two solutions :

1. a real cache eviction algorithm that's fast, efficient and easy to implement and maintain
2. a way to avoid removing entries that must be kept alive at all times

RinMaru commented:
Vexx No longer gets ingame

PatrickvL commented:
With the most recent commit, I've implemented an easier solution for now;

Instead of implementing an actual cache eviction policy, I've split up the cache in two parts.

This way, only the paletized textures get pruned (still with the poor-mans cache eviction policy : clean all), and all other resource types stay in the main cache, which implicitly prevents render target mappings getting lost.

RinMaru commented:
![Annotation 2019-11-28 183826](https://user-images.githubusercontent.com/15071652/69835276-4eda6880-120e-11ea-9b1a-76119ec33d49.jpg)
getting this error in vexx due to this PR

PatrickvL commented:
I think I'll have to redo this entire pr, in smaller increments, making sure each step doesn't regress

EternalShackles commented:
Well, latest build fixed the black textures in all games that were broken with this pr earlier
![GW OG](https://user-images.githubusercontent.com/30357118/69919553-5ef87f00-147e-11ea-8d98-b601ef716fd0.jpg)
![GW FIXED](https://user-images.githubusercontent.com/30357118/69919554-5ef87f00-147e-11ea-9609-f054b7a5e5c0.jpg)

LukeUsher commented on src/core/hle/D3D8/XbD3D8Types.h:
```diff
@@ -444,7 +444,7 @@ struct X_D3DResource
 #define X_D3DCOMMON_INTREFCOUNT_SHIFT  19
 #define X_D3DCOMMON_INTREFCOUNT_1      (1 << X_D3DCOMMON_INTREFCOUNT_SHIFT) // Dxbx addition
-#define X_D3DCOMMON_VIDEOMEMORY        0x00800000 // Not used.
+#define X_D3DCOMMON_VIDEOMEMORY        0x00000000 // Was 0x00800000, but Xbox doesn't have this flag!
```
If this flag is not present on Xbox, why not just remove the definition entirely?

PatrickvL replied:
Removing this is possible of course, but it'd also be information loss. Keeping this single line in, with the end of line comment, at least documents it clearly in code. If we ever document all of our findings somewhere that'll last, it can be removed from code, but until then I'd politely request to keep these little nuggets of information in.

PatrickvL commented:
@LastBreathToday reported :
> That latest PR on Appvoyer breaks Crash TWOC the main rooms glass floors don't render correctly when walking around plus the gates don't render anymore
> Might just be an AMD problem
> Not sure if the crystals rendering black is a AMD problem either
> Ugh AMD issue it is
> For me the floor will flicker black then normal but the elemental gates don't render and this all started from the recent PR that was causing regressions

RinMaru commented:
@PatrickvL nvr mind false alarm on AMD all is good
